src/core/utils/flash.cpp

```diff
@@ -165,55 +476,41 @@ otError Flash::Get(uint16_t aKey, int aIndex, uint8_t *aValue, uint16_t *aValueL
     return error;
 }
 
-otError Flash::Set(uint16_t aKey, const uint8_t *aValue, uint16_t aValueLength)
-{
-    return Add(aKey, true, aValue, aValueLength);
-}
-
-otError Flash::Add(uint16_t aKey, const uint8_t *aValue, uint16_t aValueLength)
+otError Flash::Add(uint16_t aKey, bool aInvalidatePrevious, const uint8_t *aValue, uint16_t aValueLength)
 {
-    bool first = (Get(aKey, 0, nullptr, nullptr) == OT_ERROR_NOT_FOUND);
-
-    return Add(aKey, first, aValue, aValueLength);
-}
-
-otError Flash::Add(uint16_t aKey, bool aFirst, const uint8_t *aValue, uint16_t aValueLength)
-{
-    otError error = OT_ERROR_NONE;
-    Record  record;
+    otError  error = OT_ERROR_NONE;
+    Record   record;
+    uint16_t size;
 
-    record.Init(aKey, aFirst);
-    record.SetData(aValue, aValueLength);
+    record.Init(aKey, aInvalidatePrevious, aValue, aValueLength);
+    size = record.GetSize();
 
-    OT_ASSERT((mSwapSize - record.GetSize()) >= kSwapMarkerSize);
+    OT_ASSERT((mSwapSize - size) >= mSwapHeaderSize);
 
-    if ((mSwapSize - record.GetSize()) < mSwapUsed)
+    if ((mSwapSize - size) < mSwapUsed)
     {
         Swap();
-        VerifyOrExit((mSwapSize - record.GetSize()) >= mSwapUsed, error = OT_ERROR_NO_BUFS);
+        VerifyOrExit((mSwapSize - size) >= mSwapUsed, error = OT_ERROR_NO_BUFS);
     }
 
-    otPlatFlashWrite(&GetInstance(), mSwapIndex, mSwapUsed, &record, record.GetSize());
+    record.Save(&GetInstance(), mSwapIndex, mSwapUsed);
```

Comment:
Should we validate integrity of record after writing? If integrity check fails, we can try append a new record?